cachi2/core/package_managers/general.py

```diff
@@ -35,6 +48,94 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
             f.write(chunk)
 
 
+async def async_download_binary_file(
+    session: aiohttp_retry.RetryClient,
+    url: str,
+    download_path: Union[Any, RootedPath],
+    auth: Optional[aiohttp.BasicAuth] = None,
+    chunk_size: int = 8192,
+) -> None:
+    """
+    Download a binary file (such as a TAR archive) from a URL using asyncio.
+
+    :param aiohttp_retry.RetryClient session: Aiohttp interface for making HTTP requests.
+    :param str url: URL for file download
+    :param str download_path: File path location
+    :param aiohttp.BasicAuth auth: Authentication for the URL
+    :param int chunk_size: Chunk size param for Response.content.read()
+    :raise FetchError: If download failed
+    """
+    try:
+        log.debug(f"Download started - {url}")
+        async with session.get(url, auth=auth, raise_for_status=True) as resp:
+            with open(download_path, "wb") as f:
+                while True:
+                    chunk = await resp.content.read(chunk_size)
+                    if not chunk:
+                        break
+                    f.write(chunk)
+
+    except Exception as exception:
+        log.error(f"Unsuccessful download: {url}")
+        # "from None" since we have the exception context in the logs
+        raise FetchError(
+            (f"exception_name: {exception.__class__.__name__}, " f"details: {exception}")
+        ) from None
+
+    log.debug(f"Download completed - {url}")
+
+
+async def async_download_files(
+    files_to_download: Dict[str, Any],
+    concurrency_limit: int,
+) -> None:
+    """Asynchronous function to download files.
+
+    :param files_to_download: Dict of files to download with file paths
+    :param concurrency_limit: Max number of concurrent tasks (downloads).
+    """
+
+    async def on_request_start(
+        session: aiohttp.ClientSession,
+        trace_config_ctx: types.SimpleNamespace,
+        params: aiohttp.TraceRequestStartParams,
+    ) -> None:
+        current_attempt = trace_config_ctx.trace_request_ctx["current_attempt"]
+        if current_attempt > 1:
+            file_name = urlparse(params.url).path.split("/")[-1]
```

Comment:
```suggestion
            file_name = params.url.path.split("/")[-1]
```
My bad, didn't realize that `params.url` was already a parsed URL

Reply:
My bad, I trust you so much that I did not question it. :)